gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/vertx/web/handler/impl/internal/mfa/MFAEnrollStep.java

```diff
@@ -47,38 +54,82 @@ public MFAEnrollStep(Handler<RoutingContext> wrapper, RuleEngine ruleEngine, Fac
     @Override
     public void execute(RoutingContext routingContext, AuthenticationFlowChain flow) {
         final Client client = routingContext.get(ConstantKeys.CLIENT_CONTEXT_KEY);
-        var context = new MfaFilterContext(routingContext, client, factorManager);
+        final MfaFilterContext context = new MfaFilterContext(routingContext, client, factorManager);
+        if (hasFactor(client, factorManager)) {
+            if (stepUp(context, client, ruleEngine)) {
+                required(routingContext, flow, context);
+            } else if (isEnrollActive(client)) {
+                switch (getEnrollSettings(client).getType()) {
+                    case OPTIONAL -> optional(routingContext, flow, context);
+                    case REQUIRED -> required(routingContext, flow, context);
+                    case CONDITIONAL -> conditional(routingContext, flow, client, context);
+                }
```

Comment:
it is throwing a null pointer error. If you want I can make a new Exception for it but I think it will not improve readability

Reply:
> it is throwing a null pointer error. If you want I can make a new Exception for it but I think it will not improve readability

No, it is OK but we have to be sure the Management API is throwing an error if the type is not valid. Should be :smile:

Reply:
There is already a test for that case